### PR TITLE
feat(java): add support for sbt projects using sbt-dependency-lock

### PR DESCRIPTION
**Pull request by stringbean**
## Description

Adds support for sbt projects using [sbt-dependency-lock](https://stringbean.github.io/sbt-dependency-lock/). This will allow scanning of Scala projects using sbt with lockfiles generated by `sbt-dependency-lock`.

## Related issues
- #2405
- Discussion: https://github.com/aquasecurity/trivy/discussions/4402
- Close https://github.com/aquasecurity/trivy/issues/6886

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Review by knqyf263 — COMMENTED** [review]
Looks awesome! Thanks for your contribution!
I have a question. Is there any way to determine if a dependency is directly used by the project or transitively installed?

File: docs/docs/coverage/language/java.md [hunk 1]
```diff
@@ -16,6 +17,7 @@ The following table provides an outline of the features Trivy offers.
 | JAR/WAR/PAR/EAR  |     Trivy Java DB     |     Include      |                  -                   |    -     |
 | pom.xml          | Maven repository [^1] |     Exclude      |                  ✓                   |  ✓[^7]   |
 | *gradle.lockfile |           -           |     Exclude      |                  ✓                   |    ✓     |
+| *.sbt.lock       |          -            |     Exclude      |                  ✓                   |    ✓     |
```
Comment (knqyf263) [inline docs/docs/coverage/language/java.md:20]: The code base doesn't seem to support the dependency graph. It needs which package depends on which package. It returns nil here.
https://github.com/aquasecurity/trivy/blob/5bb0789e1bbc44d4961542f87de4a1cd52d1fe9f/pkg/dependency/parser/sbt/lockfile/parse.go#L65

**Comment by stringbean** [comment]
> Looks awesome! Thanks for your contribution! I have a question. Is there any way to determine if a dependency is directly used by the project or transitively installed?

Not at the moment. I can look at adding that to the next major version of sbt-dependency-lock.

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+const (
+	version        = 1
+	fileNameSuffix = "sbt.lock"
+)
+
+// sbtDependencyLockAnalyzer analyzes '*.sbt.lock'
+type sbtDependencyLockAnalyzer struct {
+	logger *log.Logger
```
Comment (knqyf263) [inline pkg/fanal/analyzer/language/java/sbt/lockfile.go:31]: Seems like it's not used anywhere. We can delete it.
```suggestion
```

**Review by knqyf263 — APPROVED** [review]
LGTM.
@DmitriyLewen Can you please take a look?

File: pkg/dependency/parser/sbt/lockfile/parse.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	for _, dep := range lockfile.Dependencies {
+		libraries = append(libraries, ftypes.Package{
+			ID:        dependency.ID(ftypes.Sbt, dep.Organization+":"+dep.Name, dep.Version),
+			Name:      dep.Organization + ":" + dep.Name,
+			Version:   dep.Version,
+			Locations: []ftypes.Location{{StartLine: dep.StartLine, EndLine: dep.EndLine}},
+		})
+	}
+
+	return libraries, nil, nil
```
Comment (DmitriyLewen) [inline pkg/dependency/parser/sbt/lockfile/parse.go:65]:
```suggestion
	sort.Sort(libraries)
	return libraries, nil, nil
```
- use `var libraries ftypes.Packages` for this.
- you can remove this after that: https://github.com/aquasecurity/trivy/blob/1890227545bb7bb1f2cd31a85285a6465ba11a72/pkg/dependency/parser/sbt/lockfile/parse_test.go#L82-L90

File: pkg/dependency/parser/sbt/lockfile/parse.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	Artifacts      []sbtLockfileArtifact `json:"artifacts"`
+	Configurations []string              `json:"configurations"`
+	StartLine      int
+	EndLine        int
+}
+
+type sbtLockfileArtifact struct {
+	Name string `json:"name"`
+	Hash string `json:"hash"`
+}
```
Comment (DmitriyLewen) [inline pkg/dependency/parser/sbt/lockfile/parse.go:35]:
```suggestion
// lockfile format defined at: https://stringbean.github.io/sbt-dependency-lock/file-formats/version-1.html
type sbtLockfile struct {
	Version      int                     `json:"lockVersion"`
	Dependencies []sbtLockfileDependency `json:"dependencies"`
}

type sbtLockfileDependency struct {
	Organization   string   `json:"org"`
	Name           string   `json:"name"`
	Version        string   `json:"version"`
	Configurations []string `json:"configurations"`
	StartLine      int
	EndLine        int
}
```

I think we can avoid unmarshalling unused fields.

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	parser language.Parser
+}
+
+func newSbtDependencyLockAnalyzer(_ analyzer.AnalyzerOptions) (analyzer.PostAnalyzer, error) {
+	return &sbtDependencyLockAnalyzer{
+		parser: lockfile.NewParser(),
+	}, nil
+}
+
+func (a sbtDependencyLockAnalyzer) PostAnalyze(_ context.Context, input analyzer.PostAnalysisInput) (*analyzer.AnalysisResult, error) {
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/java/sbt/lockfile.go:39]: We don't need to parse other files (we only need lock file)
So I think it's better to use https://github.com/aquasecurity/trivy/blob/6e7f62d2debbb4acbeb7ad58ecd8ac7379f19e40/pkg/fanal/analyzer/analyzer.go#L71-L76

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		return nil, xerrors.Errorf("sbt walk error: %w", err)
+	}
+
+	return &analyzer.AnalysisResult{
+		Applications: apps,
+	}, nil
+}
+
+func (a sbtDependencyLockAnalyzer) Required(filePath string, _ os.FileInfo) bool {
+	return strings.HasSuffix(filePath, fileNameSuffix)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/java/sbt/lockfile.go:77]: Correct me, if i am wrong.

UUIC default name is `build.sbt.lock`.
But you can use any name with https://stringbean.github.io/sbt-dependency-lock/settings.html#dependencylockfile 

I think in this case we can compare file names with `build.sbt.lock`. For other names user can use https://aquasecurity.github.io/trivy/v0.52/docs/configuration/skipping/#file-patterns

Or do users often use the names `*.sbt.lock`?

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		if app == nil {
+			// no dependencies - add empty application
+			app = &types.Application{
+				Type:     types.Sbt,
+				FilePath: filePath,
+				Packages: types.Packages{},
+			}
+		}
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/java/sbt/lockfile.go:60]: We don't add empty apps.
e.g.:
https://github.com/aquasecurity/trivy/blob/6e7f62d2debbb4acbeb7ad58ecd8ac7379f19e40/pkg/fanal/analyzer/language/java/gradle/lockfile.go#L63-L65

File: pkg/fanal/analyzer/language/java/sbt/lockfile_test.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+			want: &analyzer.AnalysisResult{
+				Applications: []types.Application{
+					{
+						Type:     types.Sbt,
+						FilePath: "build.sbt.lock",
+						Packages: types.Packages{},
+					},
+				},
+			},
+		},
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/java/sbt/lockfile_test.go:33]: nit:
we try to use tests with results first (`v1 lockfile` -> `empty lockfile` in your case).

File: pkg/dependency/parser/sbt/lockfile/parse.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	Configurations []string                `json:"configurations"`
+	Dependencies   []sbtLockfileDependency `json:"dependencies"`
+}
+
+type sbtLockfileDependency struct {
+	Organization   string                `json:"org"`
+	Name           string                `json:"name"`
+	Version        string                `json:"version"`
+	Artifacts      []sbtLockfileArtifact `json:"artifacts"`
+	Configurations []string              `json:"configurations"`
```
Comment (DmitriyLewen) [inline pkg/dependency/parser/sbt/lockfile/parse.go:27]: We include only `runtime` and `compile` scopes for `pom.xml` files:
https://github.com/aquasecurity/trivy/blob/6e7f62d2debbb4acbeb7ad58ecd8ac7379f19e40/pkg/dependency/parser/java/pom/parse.go#L396-L398

I think we need to use same logic here.

**Review by DmitriyLewen — COMMENTED** [review]
Hello @stringbean 
Thanks for your work!

Left a few comments.
Take a look, please.

Regards, Dmitriy

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go (hunk 5 above)
Comment (knqyf263) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:39]: I thought we would parse other files in the future. Let's ask the future plan of sbt-dependency-lock.

Comment (stringbean) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:39]: Future improvements to sbt-dependency-lock will be added to the existing `*.sbt.lock` files.

Is it still possible to walk the directory in an `analyzer`? It is common to have multiple sub-projects in sbt and each of these would generate their own `build.sbt.lock` file.

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go (hunk 6 above)
Comment (stringbean) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:77]: I'm not aware of anyone overriding the default filename (`build.sbt.lock`) but I'm sure someone is using it! I'll change it over to use the default.

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go (hunk 5 above)
Comment (knqyf263) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:38]: What is the sub-project? Does it belong to the main project? Or are they separate projects? If you need to process those files at the same time, you need to use `PostAnalyzer`. If you only need to parse them individually, `Analyzer` is more efficient. For example, Trivy needs both `package.json` and `package-lock.json` for npm projects, and uses `PostAnalyzer` in this case.

Comment (stringbean) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:38]: sbt supports having multiple projects within a single parent project. Generally each sub-project will generate its own artifacts (JARs/WARs/GraalVM binaries etc) and will be in a sub-directory of the parent project.

Configuration for sub-projects (including dependencies) is _extremely_ flexible and can be defined in any of the following locations (or a combination):

- `build.sbt` in the sub-project dir.
- inherited from the root `build.sbt`.
- configured from the `project` directory of the root project.

The sbt-dependency-lock plugin will generate a separate `build.sbt.lock` file in each of the sub-projects that will contain all the dependencies of that sub-project.

Comment (knqyf263) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go:39]: Do you want to process `build.sbt.lock` separately or together? When processed together, all dependent packages are shown as root `build.sbt.lock`. If not, they are just different lock files and show dependencies respectively.

Comment (stringbean) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go]: After thinking about it I reckon that processing them separately is probably better for the majority of sbt users.

Comment (knqyf263) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go]: Thanks for your insight. We can use the analyzer for now and switch to the post-analyzer when we need to combine those files.

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+// sbtDependencyLockAnalyzer analyzes '*.sbt.lock'
+type sbtDependencyLockAnalyzer struct{}
+
+func (a sbtDependencyLockAnalyzer) Analyze(_ context.Context, input analyzer.AnalysisInput) (*analyzer.AnalysisResult, error) {
+	filePath := filepath.Join(input.Dir, input.FilePath)
+	parser := lockfile.NewParser()
+
+	res, err := language.Analyze(types.Sbt, filePath, input.Content, parser)
```
Comment (DmitriyLewen) [inline pkg/fanal/analyzer/language/java/sbt/lockfile.go:32]: `input.FilePath` is enough.

```suggestion
	filePath := filepath.Join(input.Dir, input.FilePath)
	parser := lockfile.NewParser()

	res, err := language.Analyze(types.Sbt, filePath, input.Content, parser)
```

File: pkg/dependency/parser/sbt/lockfile/parse.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+		return nil, nil, xerrors.Errorf("JSON decoding failed: %w", err)
+	}
+
+	var libraries ftypes.Packages
+
+	for _, dep := range lockfile.Dependencies {
+		if slices.ContainsFunc(dep.Configurations, isIncludedConfig) {
+			libraries = append(libraries, ftypes.Package{
+				ID:        dependency.ID(ftypes.Sbt, dep.Organization+":"+dep.Name, dep.Version),
+				Name:      dep.Organization + ":" + dep.Name,
```
Comment (DmitriyLewen) [inline pkg/dependency/parser/sbt/lockfile/parse.go:54]: nit:
```suggestion
			name := dep.Organization + ":" + dep.Name
			libraries = append(libraries, ftypes.Package{
				ID:        dependency.ID(ftypes.Sbt, name, dep.Version),
				Name:      name,
```

**Review by DmitriyLewen — COMMENTED** [review]
LGTM.
@stringbean left a couple of comments.

File: pkg/dependency/parser/sbt/lockfile/parse.go (hunk 11 above)
Comment (knqyf263) [reply pkg/dependency/parser/sbt/lockfile/parse.go]: Done https://github.com/aquasecurity/trivy/pull/6882/commits/3b43605b80c2a442e9a2bcf9f39f22cad2f3db19

File: pkg/fanal/analyzer/language/java/sbt/lockfile.go (hunk 10 above)
Comment (knqyf263) [reply pkg/fanal/analyzer/language/java/sbt/lockfile.go]: Done https://github.com/aquasecurity/trivy/pull/6882/commits/23ce7e672b2d25fbdad9a3bfd4e2c8cb5a2c71f8

**Pull request merged**